vim-patch:8.2.1588 - port `prompt_getprompt()`

seandewar opened the pull request:
This PR ports `prompt_getprompt()`, which is referenced in `:h prompt-buffer` since #14342 was merged.

Would appreciate feedback regarding the port of `tv_get_buf_from_arg()`.
It's also used in [v8.2.1563](https://github.com/vim/vim/commit/6f84b6db10ab86bca85e33f3fc6ee735eec8bbe5), but I haven't included that patch here.

~~I included the test for `prompt_getprompt()` in a similiar way to how `packadd_spec.lua` handles its tests.
I wasn't sure if it was better to port the test as-is, as it doesn't use any N/A stuff.~~

seandewar committed:
vim-patch:8.2.1588: cannot read back the prompt of a prompt buffer

Problem:    Cannot read back the prompt of a prompt buffer.
Solution:   Add prompt_getprompt(). (Ben Jackson, closes vim/vim#6851)
https://github.com/vim/vim/commit/077cc7aa0e0c431e97795612374fe17fe7c88803

Updated prompt_getprompt() doc to https://github.com/vim/vim/commit/cb80aa2d53e56d3aba3b3c439fb467f29a750c5e
and removed mention of method syntax usage (not supported by Nvim).

seandewar committed:
eval: port tv_get_buf_from_arg()

Cherry-picked from v8.2.1562, but uses Nvim's tv_check_str_or_nr().
Required for v8.2.1588.

It isn't used for f_bufnr() to avoid a double error message if the first
argument isn't a number or string similiar to what's seen in Vim.

janlazo commented on src/nvim/eval/funcs.c:
```diff
@@ -601,13 +601,8 @@ static void f_bufname(typval_T *argvars, typval_T *rettv, FunPtr fptr)
-    buf = tv_get_buf(&argvars[0], false);
-    emsg_off--;
+  } else if (!(buf = tv_get_buf_from_arg(&argvars[0]))) {
+    return;
```
There's no need for an early return when `buf` is set to `NULL`.

seandewar replied:
Sorry, I meant to mention that in the PR post, not just the commit message.

Btw I'm happy to address all the requested changes to this PR if it'll help. 😃

janlazo commented on src/nvim/eval/funcs.c:
```diff
@@ -627,6 +622,9 @@ static void f_bufnr(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   } else {
+    // Don't use tv_get_buf_from_arg(); we continue if the buffer wasn't found
+    // and the second argument isn't zero, but we want to return early if the
+    // first argument isn't a string or number so only one error is shown.
```
Is this because patch v8.2.1183 is not ported yet?

seandewar replied:
I don't think so as Vim still has the behaviour. If you try `bufnr()` with a first argument that's not a string/number:
```vim
:echo bufnr([], 1)
```

In Vim you get:
```
E745: Using a List as a Number
E730: Using List as a String
-1
```

In Nvim you get:
```
E745: Expected a Number or a String, List found
-1
```

janlazo commented on src/nvim/eval/funcs.c:
```diff
@@ -6491,6 +6491,26 @@ static void f_prompt_setinterrupt(typval_T *argvars,
 
+/// "prompt_getprompt({buffer})" function
+void f_prompt_getprompt(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  FUNC_ATTR_NONNULL_ALL
```
Weird that this didn't cause a compile error. It failed on my machine last time I tried to add function attribute on eval functions.

seandewar replied:
Some funcs like `f_luaeval()`, `f_reltimestr()` have attribs so I hope things'll be OK. :smile:

janlazo commented on test/functional/legacy/prompt_buffer_spec.lua:
```diff
@@ -4,150 +4,203 @@ local feed= helpers.feed
+  describe('(legacy test)', function()
+    before_each(function()
+      source([=[
+        func Test_prompt_buffer_getbufinfo()
```
In the long run, it would be easier if you ported the entire `Test_prompt_buffer.vim` and modified it to run this test only. I don't mind if you literred most of the test cases there with `throw "skipped: TODO"`.

janlazo reviewed (approved):
It's good enough. Thanks.

seandewar committed:
oldtest: port test_prompt_buffer.vim

Included from before v8.2.1588. Required for v8.2.1588.
Many tests use stuff like term_sendkeys() which is N/A, so are skipped.

seandewar commented:
I removed the early return & ported `test_prompt_buffer.vim` with tests skipped except `prompt_getprompt()`'s.
`prompt_buffer_spec.lua` changes were reverted.

[Diff between the previous version of this PR is here](https://github.com/neovim/neovim/compare/04c0f302eba71297bd4ca216829d86169fe5abc8..1bee8d3a81bf481cab6d924a71082b7d2d109030).

janlazo commented on src/nvim/testdir/test_prompt_buffer.vim:
```diff
@@ -0,0 +1,194 @@
+func CanTestPromptBuffer()
+  " We need to use a terminal window to be able to feed keys without leaving
+  " Insert mode.
+  CheckFeature terminal
```
Comment out.

janlazo commented:
> Looks like that pattern should be something like [0-2] seconds\= ago instead?

Sure.

seandewar committed:
fix(oldtest): fix assert_match() for :buffers t

Previous pattern could cause test to fail if ":buffers t" reported
"1 second ago".

janlazo commented:
Perfect.

@seandewar Thank you for your contribution.
